### PR TITLE
Implement version checking for JITaaS

### PR DESCRIPTION
**Pull request by harryyu1994**
Newly added exceptions:
1. **StreamMessageTypeMismatch**: a server side exception, the throw happens in `read()` and `readCompileRequest()` when the received message from client is not of the expected type.
When we catch this exception, we terminate the compilation without closing the connection by writing to client `stream->finishCompilation(compilationStreamMessageTypeMismatch)` with the error code `compilationStreamMessageTypeMismatch`.
2. **StreamVersionIncompatible**: the throw happens in `readCompileRequest()` when client version is not compatible with server version. Used for both client and server, but mainly for server. When we catch this exception, we terminate the compilation by writing to client `stream->finishCompilation(compilationStreamVersionIncompatible)` with the error code `compilationStreamVersionIncompatible`

Newly added compilation error codes: The error codes are handled in routine `shouldRetryCompilation` on the client upon receiving the `stream->finishCompilation()` message from server.
1. **compilationStreamMessageTypeMismatch**: tryCompilingAgain = true
2. **compilationStreamVersionIncompatible**: we mark the server as incompatible and proceed with recompilation locally.

Version Checking:
3 status: INCOMPATIBLE, COMPATIBLE, NOT_DONE
Client message now has 3 fields: type, data and version
Server message now has 2 fields: type and data
The version field is an uint64_t flag: CONFIGURATION_FLAG (32bit) + MAJOR_NUMBER(8bit) + MINOR_NUMBER(16bit) + PATCH_NUMBER(8bit)

Incorporated the version checking mechanism in `buildCompileRequest` and `readCompileRequest`:
1. **buildCompileRequest**: a client side routine, if status is NOT_DONE, we set the version before sending for the server to check for compatibility. If status is COMPATIBLE, that means we have already confirmed that server is compatible, so we want to send data without the version field to bypass the check. If status is INCOMPATIBLE, it means server has already rejected the client request, so we should always do local compilations instead.

2. **readCompileRequest**: a server side routine, if version is not 0 (default is 0), then it means client wants to check for compatibility. If version is 0, it means that the client has already been checked and it's compatible.

[skip ci]
Issue: #4467

Signed-off-by: Harry Yu <harryyu1994@gmail.com>

### CODE REVIEW

**Comment by harryyu1994** [comment]
Will reopen this PR once I have the new code ready, abandoning everything in here.

**Comment by harryyu1994** [comment]
@mpirvu ran some basic tests. 
1. a compressedref server with a non-compressedref client and visa versa, client ends up compiling locally.
2. server/client with different versions, client ends up compiling locally
3. server/client with same versions, normal operations

Ready for review.

**Comment by harryyu1994** [comment]
Newly added exceptions:
1. **StreamMessageTypeMismatch**: a server side exception, the throw happens in `read()` and `readCompileRequest()` when the received message from client is not of the expected type.
When we catch this exception, we terminate the compilation without closing the connection by writing to client `stream->finishCompilation(compilationStreamMessageTypeMismatch)` with the error code `compilationStreamMessageTypeMismatch`.
2. **StreamVersionIncompatible**: the throw happens in `readCompileRequest()` when client version is not compatible with server version. Used for both client and server, but mainly for server. When we catch this exception, we terminate the compilation by writing to client `stream->finishCompilation(compilationStreamVersionIncompatible)` with the error code `compilationStreamVersionIncompatible`

Newly added compilation error codes: The error codes are handled in routine `shouldRetryCompilation` on the client upon receiving the `stream->finishCompilation()` message from server.
1. **compilationStreamMessageTypeMismatch**: tryCompilingAgain = true
2. **compilationStreamVersionIncompatible**: we mark the server as incompatible and proceed with recompilation locally.

Version Checking:
3 status: INCOMPATIBLE, COMPATIBLE, NOT_DONE
Client message now has 3 fields: type, data and version
Server message now has 2 fields: type and data
The version field is an uint64_t flag: CONFIGURATION_FLAG (32bit) + MAJOR_NUMBER(8bit) + MINOR_NUMBER(16bit) + PATCH_NUMBER(8bit)

Incorporated the version checking mechanism in `buildCompileRequest` and `readCompileRequest`:
1. **buildCompileRequest**: a client side routine, if status is NOT_DONE, we set the version before sending for the server to check for compatibility. If status is COMPATIBLE, that means we have already confirmed that server is compatible, so we want to send data without the version field to bypass the check. If status is INCOMPATIBLE, it means server has already rejected the client request, so we should always do local compilations instead.

2. **readCompileRequest**: a server side routine, if version is not 0 (default is 0), then it means client wants to check for compatibility. If version is 0, it means that the client has already been checked and it's compatible.

**Comment by harryyu1994** [comment]
@mpirvu Hi Marius, this PR is ready for review. In the meantime I will try various tests on it.

File: runtime/compiler/control/CompilationThread.cpp [hunk 1]
```diff
@@ -7036,7 +7047,7 @@ TR::CompilationInfoPerThreadBase::preCompilationTasks(J9VMThread * vmThread,
                }
             }
 
-         if (!doLocalCompilation)
+         if (!doLocalCompilation && JITaaS::J9ClientStream::getVersionCheckStatus() != JITaaS::INCOMPATIBLE)
```
Comment (mpirvu) [inline runtime/compiler/control/CompilationThread.cpp]: If logic is added to `shouldPerformLOcalComp()` to check for the version, then this code is not needed.

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 2]
```diff
@@ -242,9 +242,6 @@ bool handleServerMessage(JITaaS::J9ClientStream *client, TR_J9VM *fe)
    // If JVM has unloaded classes inform the server to abort this compilation
    if (compInfoPT->compilationShouldBeInterrupted())
       {
-      if (response != J9ServerMessageType::compilationCode)
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: We should still send a message to the server to abort current compilation. Otherwise the server will hit a "message not expected" scenario.

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 3]
```diff
@@ -2479,16 +2476,10 @@ remoteCompile(
          }
       catch (const JITaaS::StreamFailure &e)
          {
-         if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: Why were these lines deleted?

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 4]
```diff
@@ -2504,16 +2495,10 @@ remoteCompile(
             }
          catch (const JITaaS::StreamFailure &e)
             {
-            if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: Same here: why don't we want these messages?

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 5]
```diff
@@ -2579,7 +2564,6 @@ remoteCompile(
          auto comp = compInfoPT->getCompilation();
          if (TR::Options::getVerboseOption(TR_VerboseCompilationDispatch))
             TR_VerboseLog::writeLineLocked(TR_Vlog_DISPATCH, "Interrupting remote compilation of %s @ %s", comp->signature(), comp->getHotnessName());
-         client->writeError();
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: Like in other cases, we need a specific message to let the server know the client is not going to continue with this compilation.

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 6]
```diff
@@ -2603,19 +2587,27 @@ remoteCompile(
       resolvedMirrorMethodsPersistIPInfo = std::get<7>(recv);
       if (statusCode >= compilationMaxError)
          throw JITaaS::StreamTypeMismatch("Did not receive a valid TR_CompilationErrorCode as the final message on the stream.");
+      if (statusCode != compilationStreamVersionIncompatible)
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: We could store this in the stream itself. As long as the stream is not recreated we have the version check already done.

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 7]
```diff
@@ -2603,19 +2587,27 @@ remoteCompile(
...
-         TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, e.what());
-         
       client->~J9ClientStream();
       TR_Memory::jitPersistentFree(client);
       compInfoPT->setClientStream(NULL);
 
       compiler->failCompilation<JITaaS::StreamFailure>(e.what());
       }
+   catch (const JITaaS::StreamVersionIncompatible &e)
+      {
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: Disconnecting the incompatible stream is the right thing to do. We just have to make sure the client does not try over and over to connect to an incompatible server.

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 8]
```diff
@@ -4277,7 +4269,7 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
       if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
          TR_VerboseLog::writeLineLocked(TR_Vlog_JITaaS, "Stream failed while compThreadID=%d was reading the compilation request: %s", 
             getCompThreadId(), e.what());
-      stream->cancel(); // This does nothing for raw sockets
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: If `cancel()` does nothing, we should delete the function altogether.

File: runtime/compiler/rpc/protos/compile.proto [hunk 9]
```diff
@@ -51,6 +51,7 @@ enum J9ServerMessageType
    mirrorResolvedJ9Method = 1;
    get_params_to_construct_TR_j9method = 2;
    getUnloadedClassRanges = 3;
+   compilationRequest = 4;
```
[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/protos/compile.proto]: A new message type on the server side.
The initial client remote compilation requests all have this type.

File: runtime/compiler/rpc/J9Server.hpp [hunk 10]
```diff
@@ -57,12 +57,31 @@ class J9ServerStream : J9Stream
...
+
+   template <typename... T>
+   std::tuple<T...> readCompileRequest()
+      {
+      readBlocking(_cMsg);
+      if (_cMsg.type() != J9ServerMessageType::compilationRequest)
+         {
+         throw StreamMessageTypeMismatch(J9ServerMessageType::compilationRequest, _cMsg.type());
+         }
+      if (_cMsg.version() != 0 && _cMsg.version() != ((uint64_t) CONFIGURATION_FLAG << 32 | MAJOR_NUMBER << 24 | MINOR_NUMBER << 8))
```
[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/J9Server.hpp]: I thought configuration flags should be a perfect match.

I think we have discussed this that the client's processor features don't have to match the server's. Server will generate code based on client's processor features. So these things shouldn't even be on the configuration flags.

File: runtime/compiler/control/JITaaSCompilationThread.cpp (hunk 7 above)
Comment (harryyu1994) [reply runtime/compiler/control/JITaaSCompilationThread.cpp]: the reason i used a static status for compatibility is to make sure the client does not try to connect over and over again, I failed to see how that can be achieved with a non static compatibility status.

**Comment by harryyu1994** [comment]
Original plan:
when I had a static Version Checking Status,
as soon as one of the compilation requests encounters a version compatibility failure, the client just switches to local compilation and never looks back.
I agree that this original plan does not take into consideration that the server might change during the run of a client.

New plan:
we need a static counter to count the number of version compatibility failures. Say if we exceed 5, then we switch to local compilation. After that we periodically check if the server version has changed.

Do you agree with the new plan?

**Comment by harryyu1994** [comment]
Another comment I had,
For the configuration flags, I think that they should be a perfect match for compatibility checking.

I think we have discussed this that the client's processor features don't have to match the server's. Server will generate code based on client's processor features. So these things shouldn't even be included in the configuration_flags.

**Comment by mpirvu** [comment]
I agree that processor configuration flags should be separate from the "configuration"

I also agree with the plan to check from time to time to see if the version compatibility failure has been solved by connecting to a different server.

**Comment by harryyu1994** [comment]
Ready for review.

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 11]
```diff
@@ -2603,19 +2591,28 @@ remoteCompile(
       resolvedMirrorMethodsPersistIPInfo = std::get<7>(recv);
       if (statusCode >= compilationMaxError)
          throw JITaaS::StreamTypeMismatch("Did not receive a valid TR_CompilationErrorCode as the final message on the stream.");
+      if (statusCode == compilationStreamVersionIncompatible)
+         throw JITaaS::StreamVersionIncompatible();
+      client->setVersionCheckStatus();
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: Is this the only place where we need to set the status of the version check? At this point the entire compilation is finished, so the next compilation on the same stream will be fine, but how about all the other messages sent for this compilation? Will they have the version field set to 0 to avoid message bloat?

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 12]
```diff
@@ -4653,3 +4657,22 @@ TR::CompilationInfoPerThreadRemote::clearPerCompilationCaches()
    clearPerCompilationCache(_resolvedMirrorMethodsPersistIPInfo);
    clearPerCompilationCache(_classOfStaticMap);
    }
+
+void
+TR::CompilationInfoPerThreadRemote::terminateClient(uint64_t clientId, TR::CompilationInfo* compInfo, J9VMThread* compThread)
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: I feel that 'deleteClientSessionData' is more in line with the function actually does.

File: runtime/compiler/rpc/protos/compile.proto [hunk 13]
```diff
@@ -300,8 +303,10 @@ message J9ServerMessage
 
 message J9ClientMessage
    {
-   bool status = 1;
+   J9ServerMessageType type = 1;
    AnyData data = 2;
+   uint64 version = 3;
+   uint64 clientid = 4;
```
[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/protos/compile.proto]: i guess we can continue to send it using `AnyData data` and not introducing a new type.

File: runtime/compiler/rpc/protos/compile.proto [hunk 14]
```diff
@@ -300,8 +303,10 @@ message J9ServerMessage
 
 message J9ClientMessage
    {
-   bool status = 1;
+   J9ServerMessageType type = 1;
```
[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/protos/compile.proto]: There is a ton of `J9ServerMessageType`s, can I do the name change in another PR so that it doesn't hide some of the key things in this change.

[the remark this answers is not in the document]
Comment (mpirvu) [reply runtime/compiler/rpc/protos/compile.proto]: Agreed

File: runtime/compiler/rpc/J9Stream.hpp [hunk 15]
```diff
@@ -23,11 +23,21 @@
 #ifndef J9_STREAM_HPP
 #define J9_STREAM_HPP
 
+#include <google/protobuf/io/zero_copy_stream_impl.h>
```
[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/J9Stream.hpp]: I moved it here because apparently `J9Stream.hpp` depends on these 2 headers. I found that out when I tried to include only `J9Stream.hpp` in `J9Stream.cpp`. So due to that `J9Stream.cpp`, `J9Client.cpp` and `J9Server.cpp` all need `J9Stream.hpp` and `J9Stream.hpp` needs `<google/protobuf/io/zero_copy_stream_impl.h>`, I put them in `J9Stream.hpp`. If this is not the correct practice, then I'll move them back to `J9Client.cpp`, `J9Server.cpp` and newly created `J9Stream.cpp`.

File: runtime/compiler/rpc/J9Client.hpp [hunk 16]
```diff
@@ -82,11 +94,60 @@ class J9ClientStream : J9Stream
 
    void shutdown();
 
+   void writeError(uint64_t clientId=0)
+      {
+      _cMsg.set_type(J9ServerMessageType::compilationAbort);
+      _cMsg.mutable_data()->clear_data();
```
[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/J9Client.hpp]: it does not, it only clears `Anydata data`

[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/J9Client.hpp]: but I've moved `set_type` after `clear_data()` anyways

File: runtime/compiler/rpc/StreamTypes.hpp [hunk 17]
```diff
@@ -48,7 +52,16 @@ class StreamFailure: public virtual std::exception
 class StreamCancel: public virtual std::exception
    {
 public:
-   virtual const char* what() const throw() { return "compilation canceled by client"; }
+   StreamCancel(J9ServerMessageType type, uint64_t clientId=0) : _type(type), _clientId(clientId) { }
```
[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/StreamTypes.hpp]: `_clientID` is used for `terminateClient()` in a catch block.

File: runtime/compiler/control/JITaaSCompilationThread.cpp (hunk 11 above)
Comment (harryyu1994) [reply runtime/compiler/control/JITaaSCompilationThread.cpp]: `buildCompileRequest` sends version, and only the `buildCompileRequest` operation sends version. Regular `write` operation won't send version info.
A summary of how it operates:
`readCompileRequest` receives version and check, if fails, throws `StreamVersionIncompatible`, then server terminates the compilation and sends back a `compilationCode` message with `statusCode` being set to `compilationStreamVersionIncompatible`, client does not send any other messages to the server for this compilation if the version check failed. If pass, all remaining write operations do not send version.

File: runtime/compiler/rpc/StreamTypes.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (harryyu1994) [reply runtime/compiler/rpc/StreamTypes.hpp]: Type is to indicate that there are two types of `StreamCancel`:
1. client aborts compilation
2. client is exiting
We handle these two types differently, not only used to decide what message to print.
I put them both under `StreamCancel` because they are both under the category where client cancels the compilation.

**Comment by harryyu1994** [comment]
@mpirvu Hi Marius, this change is ready for review. Passed all preliminary manual tests. Going to try with docker containers next.

File: runtime/compiler/control/JITaaSCompilationThread.cpp (hunk 11 above)
Comment (harryyu1994) [reply runtime/compiler/control/JITaaSCompilationThread.cpp]: In `buildCompileRequest`:
we clear the version after we send the message, so that proceeding `write` operations won't have version set.
```
      if (getVersionCheckStatus() == NOT_DONE)
         {
         _cMsg.set_version(getJITaaSVersion());
         write(J9ServerMessageType::compilationRequest, args...);
         _cMsg.clear_version();
         }
```

File: runtime/compiler/rpc/J9Client.hpp [hunk 18]
```diff
@@ -23,8 +23,6 @@
 #ifndef J9_CLIENT_H
 #define J9_CLIENT_H
 
-#include <google/protobuf/io/zero_copy_stream_impl.h>
-#include "compile/CompilationTypes.hpp"
 #include "rpc/StreamTypes.hpp"
 #include "rpc/ProtobufTypeConvert.hpp"
 #include "j9.h"
```
Comment (mpirvu) [inline runtime/compiler/rpc/J9Client.hpp]: Is it possible to remove this `j9.h` or include something directly from the OMR layer? The goal is to contribute this communications layer to OMR, so we want to eliminate anything that is J9 related.

File: runtime/compiler/rpc/J9Client.hpp [hunk 19]
```diff
@@ -82,11 +95,58 @@ class J9ClientStream : J9Stream
...
+      }
+
+   static void incrementIncompatibilityCount(OMRPortLibrary *portLibrary)
+      {
+      OMRPORT_ACCESS_FROM_OMRPORT(portLibrary);
+      _incompatibilityCount += 1;
+      _incompatibleStartTime = omrtime_current_time_millis();
+      }
+
+   static uint32_t getIncompatibilityCount(OMRPortLibrary *portLibrary)
```
Comment (mpirvu) [inline runtime/compiler/rpc/J9Client.hpp]: I see that this function is only used in `isServerCompatible()` below. Maybe include its body into `isServerCompatible()` and remove `getCompatibilityCount()`. Also, please add some comments describing the intention of the code.

File: runtime/compiler/rpc/J9Client.hpp [hunk 20]
```diff
@@ -82,11 +95,58 @@ class J9ClientStream : J9Stream
...
    void shutdown();
 
+   template <typename ...T>
+   void writeError(J9ServerMessageType type, T... args)
+      {
+      _cMsg.set_type(type);
+      if (type == J9ServerMessageType::compilationAbort)
+         _cMsg.mutable_data()->clear_data();
+      else
+         setArgs<T...>(_cMsg.mutable_data(), args...);
```
Comment (mpirvu) [inline runtime/compiler/rpc/J9Client.hpp]: Question: for clientTerminate we only send clientID. Are all the other fields clear or do we have to clear them first and then do `setArgs`?

File: runtime/compiler/control/CompilationThread.cpp [hunk 21]
```diff
@@ -6781,16 +6768,17 @@ bool isMethodIneligibleForAot(J9Method *method)
...
-         (TR::Options::getCmdLineOptions()->getOption(TR_EnableJITaaSHeuristics) || localColdCompilations))
-         doLocalComp = true;
+   bool doLocalComp = false;
+   static char *localColdCompilations = feGetEnv("TR_LocalColdCompilations");
+   // As a heuristic, cold compilations should be performed locally because
+   // they are supposed to be cheap with respect to memory and CPU.
+   //
+   if (entry->_optimizationPlan->getOptLevel() <= cold &&
+      (TR::Options::getCmdLineOptions()->getOption(TR_EnableJITaaSHeuristics) || localColdCompilations) || 
+      !JITaaS::J9ClientStream::isServerCompatible(OMRPORT_FROM_J9PORT(_jitConfig->javaVM->portLibrary)))
```
Comment (mpirvu) [inline runtime/compiler/control/CompilationThread.cpp]: Don't you have to pass just the portLibrary here? `_jitConfig=>javaVM->portLibrary` instead of `OMR_PORT_FROM_J9PORT...`

File: runtime/compiler/control/JITaaSCompilationThread.cpp [hunk 22]
```diff
@@ -4064,7 +4061,18 @@ TR::CompilationInfoPerThreadRemote::processEntry(TR_MethodToBeCompiled &entry, J
...
    try
       {
-      auto req = stream->read<uint64_t, uint32_t, J9Method *, J9Class*, TR_Hotness, std::string,
+      auto response = stream->readCompileRequest();
+      if (response == JITaaS::J9ServerMessageType::clientTerminate)
+         {
+         clientId = std::get<0>(stream->getRecvData<uint64_t>());
+         throw JITaaS::StreamCancel(response, clientId);
+         }
+      else if (response != JITaaS::J9ServerMessageType::compilationRequest)
```
Comment (mpirvu) [inline runtime/compiler/control/JITaaSCompilationThread.cpp]: I am thinking that this check should be implemented inside `readCompileRequest()` much like `read()` does. We already have a catch block for `StreamMessageTypeMissmatch`

File: runtime/compiler/control/CompilationThread.cpp (hunk 21 above)
Comment (harryyu1994) [reply runtime/compiler/control/CompilationThread.cpp]: I thought we have a `J9PortLibrary` and a `OMRPortLibrary`
and _jitConfig->javaVM->portLibrary gives you the J9PortLibrary
and to get OMRPortLibrary, we need to use `OMRPORT_FROM_J9PORT`. Correct me if I'm wrong.

Comment (mpirvu) [reply runtime/compiler/control/CompilationThread.cpp]: You may be right. In that case I am ok with the code.

File: runtime/compiler/rpc/J9Client.hpp (hunk 20 above)
Comment (harryyu1994) [reply runtime/compiler/rpc/J9Client.hpp]: `setArgs` will overwrite things that were previously in `_cMsg.mutable_data()`, there is no need to first clear them.

Comment (harryyu1994) [reply runtime/compiler/rpc/J9Client.hpp]: `mutable_data()` will be re-constructed based on the `T... args` we pass in.

**Comment by harryyu1994** [comment]
All concerns addressed. I'm going to redo my tests now. Ready for a final review.

**Comment by harryyu1994** [comment]
Tests working as expected.

I tried with my acmeair containers:
2 servers with different version
1 client

Start client with incompatible server
Client performs local compilations and check back with server every 10 seconds

Now replace incompatible server with compatible server
Client switches to remote compilation when it performs the next check
